cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```diff
@@ -440,19 +440,15 @@ public void stop(Id.Program programId) throws Exception {
 
   private void stopProgramIfRunning(Id.Program programId)
     throws InterruptedException, ExecutionException {
-    ProgramRuntimeService.RuntimeInfo programRunInfo = findRuntimeInfo(programId.getNamespaceId(),
-                                                                       programId.getApplicationId(),
-                                                                       programId.getId(),
-                                                                       programId.getType(),
-                                                                       runtimeService);
+    ProgramRuntimeService.RuntimeInfo programRunInfo = findRuntimeInfo(programId, runtimeService);
     if (programRunInfo != null) {
-      doStop(programRunInfo);
+      doStop(programRunInfo, programId);
     }
   }
 
-  private void doStop(ProgramRuntimeService.RuntimeInfo runtimeInfo)
-    throws ExecutionException, InterruptedException {
-    Preconditions.checkNotNull(runtimeInfo, UserMessages.getMessage(UserErrors.RUNTIME_INFO_NOT_FOUND));
+  private void doStop(ProgramRuntimeService.RuntimeInfo runtimeInfo,
+                      Id.Program programId) throws ExecutionException, InterruptedException {
+    Preconditions.checkNotNull(runtimeInfo, UserMessages.getMessage(UserErrors.RUNTIME_INFO_NOT_FOUND), programId);
```

Comment:
When the program is already stopped. The message that we return to the user would now be: 
``Unable to retrieve runtime information for <programId>``

This is not a good message to the user. 

We should say:
``Unable to stop program <programId>. The program is not running.``

Also, what is the HTTP code we will return in this case? It should not be 500 (Server error) but should be 409 (User error: Conflict)